include/arch/gdbstub.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ARCH_GDBSTUB_H_
+#define ZEPHYR_INCLUDE_ARCH_GDBSTUB_H_
+
+#include <inttypes.h>
+
+#if defined(CONFIG_X86)
+#include <arch/x86/ia32/gdbstub.h>
+#endif
+
+/**
+ * ARCH_GDB_NUM_REGISTERS is architecure specific and
+ * this symbol must be defined in architecure specific header
+ */
+#define GDB_NUM_REGISTERS ARCH_GDB_NUM_REGISTERS
+
+/**
+ * @brief Architecture layer debug start
+ *
+ * This function is called by @c z_gdb_start()
+ */
+void arch_gdb_start(void);
+
+/**
+ * @brief Read a byte from given memory address
+ *
+ * @param addr Memory address to be read
+ * @param val [out] Memory contents
+ */
+void arch_gdb_mem_readb(uintptr_t addr, uint8_t *val);
```

Comment:
quick question: under what circumstances do we need this abstraction? the current implementations just read/write memory which we could just do directly from kernel code

Reply:
not now, but when we have virtual memory and debugging user threads we will need to translations afaik. I know that a lot of things will need to adapt to support it but at least the interfaces I think won't be necessary.

Reply:
do you have some additional detail about this?

my plans in this area were, if we look at memory in a GDB session that triggers a pagefault due to a swapped page, to allow the pagefault handler to do its thing and then retry. all of this should be transparent. maybe there's something I'm not considering? examples in other OS implementations might be helpful

Reply:
that should work, will remove it from arch layer and if need we change it later.

Reply:
the current plan for virtual memory is a single view of RAM for the entire system from the kernel's perspective, and not independent address spaces, so we should be good with how this is currently.